### PR TITLE
React i18next - with translation strings extraction

### PR DESCRIPTION
**Pull request by pettinarip**
This PR reimplements the previously reverted PR #8406 for memory consumption issues on Gatsby Cloud.

## Description

All the work done in this PR is coming from: #9261

## Related Issue

#8721
#8345

### CODE REVIEW

File: docs/locales-process.md
Comment (pettinarip) [inline docs/locales-process.md:9]: A bit of context on how the whole process works. Would appreciate any feedback on this.

Comment (pettinarip) [reply docs/locales-process.md:9]: Also a recommended thing to read first that will help you understand and review the PR.

File: src/scripts/createLocales.ts
Comment (pettinarip) [inline src/scripts/createLocales.ts:1]: Main difference between this PR and the previous one #8406

File: docs/locales-process.md
Comment (samajammin) [inline docs/locales-process.md:6]: Diagrams for the win! 😍

File: src/components/ButtonDropdown.tsx
Comment (corwintines) [inline src/components/ButtonDropdown.tsx:122]: This isnt used anywhere now, I think we can just remove this from the file altogether

File: src/components/Roadmap.tsx
Comment (corwintines) [inline src/components/Roadmap.tsx:3]:
```suggestion
```
Not needed anymore

File: src/components/Footer.tsx
Comment (wackerow) [inline src/components/Footer.tsx:66]: Do we want to apply this here?
Line 342 below passes this as a key to the Translation component:
```ts
            {linkSections.map((section: LinkSection, idx) => (
              <Box key={idx}>
                <Heading as="h3" fontSize="sm" lineHeight="1.6" my="1.14em">
                  <Translation id={section.title} />
                </Heading>
                <List fontSize="sm" lineHeight="1.6" fontWeight="400" m={0}>
                  {section.links.map((link, linkIdx) => (
```
@pettinarip Should we remove the use of the `<Translation/>` component below?

File: src/components/Nav/index.tsx
Comment (wackerow) [inline src/components/Nav/index.tsx:337]: Line 433 below, we map this as a key... similar to a note elsewhere, should we remove the use of `<Tranlation/>` below?
```ts
          {ednLinks.map((link, idx) => (
            <NavLink
              key={idx}
              to={link.to}
              isPartiallyActive={link.isPartiallyActive}
            >
              <Translation id={link.text} />
            </NavLink>
          ))}
```

File: src/scripts/createLocales.ts
Comment (wackerow) [inline src/scripts/createLocales.ts:83]: @pettinarip Curious if there is a reason we're using "Lng" without the "a" for language in this script... we don't do this anywhere else in the codebase.

**Review by wackerow — COMMENTED** [review]
This looks great @pettinarip!! Well done! 🎉 
Left a couple comments, but aside from this few small things this seems to be working great for me on my end.

**Comment by minimalsm** [comment]
Bug: When I'm on the tutorials homepage in a non-English language, if I click on a tutorial, it prepends an additional lang path, causing a 404. Confirmed broken in French and Chinese, but works as expected in English.

So, instead of:

[/zh/developers/tutorials/run-node-raspberry-pi/](https://ethereumorgwebsitedev01-i18nextractstrings.gatsbyjs.io/zh/developers/tutorials/run-node-raspberry-pi/)

I get:

[/zh/zh/developers/tutorials/run-node-raspberry-pi/](https://ethereumorgwebsitedev01-i18nextractstrings.gatsbyjs.io/zh/zh/developers/tutorials/run-node-raspberry-pi/)

**Comment by minimalsm** [comment]
We don't have any highlighting on active breadcrumbs. 

Note this sometimes happens on Gatsby preview deploys so might be nothing to do with this PR.

Screenshot 2023-01-05 at 12 55 49

File: src/pages/developers/tutorials.tsx
Comment (minimalsm) [inline src/pages/developers/tutorials.tsx:428]: Not something to fix here, but noting we should require tutorials to have published dates on build.

File: src/components/ButtonDropdown.tsx
Comment (pettinarip) [reply src/components/ButtonDropdown.tsx:138]: Oh I guess that's why I had that import unused...I forgot to wrap this with the `t` function. Thanks!

File: src/components/Footer.tsx
Comment (pettinarip) [reply src/components/Footer.tsx:66]: Good catch. Yea, we don't need the `<Translation />` component here since we are already doing the translations above.

The reason I've changed it is because the scanner we are using can't detect translation keys from variables. In this case, the scanner will find this line `<Translation id={section.title} />` and it will not know what key `section.title` holds. Ideally, we should wrap all the translation key strings with the `t` function or the `Translation` component.

File: src/components/Nav/index.tsx
Comment (pettinarip) [reply src/components/Nav/index.tsx:337]: Yes, same case as the prev comment. Thanks.

File: src/scripts/createLocales.ts
Comment (pettinarip) [reply src/scripts/createLocales.ts:83]: Good catch. Didn't realize that. I guess I got confused with the terminology used by the scanner's code 🤦🏼 Will fix that.

File: src/components/ButtonDropdown.tsx
Comment (pettinarip) [reply src/components/ButtonDropdown.tsx:138]: Oh no, now I remember. So, `text` now is not more a `TranslationKey`, its a `string`. Bc the translation is done by the consumer. Here we are expecting to receive an already translated string.

For the same reason I explained in the other comment, the scanner can't extract the translation keys from variables.

**Comment by pettinarip** [comment]
As described in the PR description, the goal of this PR was to reduce memory consumption. At the moment of creation of the PR, it was building fine but after adding a few tweaks due to the code review, we started having memory issues again.

This means that the PR, at the moment of creation, was still consuming a lot of memory. To merge this PR, we want to be confident that we are far away from exceeding the memory limits and that we are not going to have future problems like this again.

I see 2 paths to go from here:

- Abandon the PR and the effort since we are planning on moving to NextJS in the near future. NextJS implementation can fix these problems that are directly related to how Gatsby's build system is designed
- Put the PR on hold and keep investigating what we can do to optimize the memory consumption and merge this PR later when we are in a good spot

Having said that, since the end goal of the i18n lib change is reducing the final JS bundles (which directly affect the UX), I think we should put this on hold and keep investing time in figuring out how to solve this problem.

There are other parts of the build process that may affect memory consumption as described [here](https://www.gatsbyjs.com/docs/how-to/performance/resolving-out-of-memory-issues/).

1. One major thing to optimize is image processing. There was [a PR](https://github.com/ethereum/ethereum-org-website/pull/6585) where we were trying to do this. We could recover that work and apply it to see how it impacts the overall memory consumption
2. Another thing is to [upgrade Gatsby to v5](https://github.com/ethereum/ethereum-org-website/issues/8586) since they say there are many optimizations implemented in this version
3. Keep investigating how we can break even more out the translations namespaces to reduce the amount of text we inject into bundles

Currently, Gatsby Cloud is [offering 8gb](https://support.gatsbyjs.com/hc/en-us/articles/360052326714-Platform-Limits) on our current plan. That should be enough. I understand that Vercel is also offering the same memory limits.

cc @samajammin @minimalsm @corwintines @wackerow

**Comment by samajammin** [comment]
Thanks for the update @pettinarip!

> since the end goal of the i18n lib change is reducing the final JS bundles (which directly affect the UX), I think we should put this on hold and keep investing time in figuring out how to solve this problem

I agree that is what we ultimately should focus on solving, the JS bundle is the user-facing issue (vs. reducing memory consumption on builds).

I think all those ideas are decent options to explore. I would be curious to see a quick spike on upgrading to Gatsby v5 & seeing if that fixes any of our issues 🤔 if not, perhaps images are an area to optimize like you mention. Realistically I'm not sure we should say "let's just solve it with Next.js" as that feels like a massive undertaking & I'm not sure if that makes sense as a near-term priority... but if this issue persists & we don't see other options, perhaps it's an argument to prioritize it.

One more alternative to explore would be simply ask Gatsby how much to increase memory limits. It's certainly a bandaid solution but if it doesn't add much additional cost it might be the most reasonable approach in the short term.

File: src/components/Layer2/Layer2Onboard.tsx
Comment (minimalsm) [inline src/components/Layer2/Layer2Onboard.tsx:271]: .

File: gatsby-ssr.tsx
Comment (corwintines) [inline gatsby-ssr.tsx:6]: I think we can just remove this file. I dont think this is doing anything anymore, unless there is a reason we need to keep it around for gatsby

**Review by corwintines — CHANGES_REQUESTED** [review]
Just noticed this, otherwise I think we are good 🚀

**Review by corwintines — APPROVED** [review]
🚀

**Pull request merged**
